ArjunaJTA/jta/classes/com/arjuna/ats/jta/common/JTAEnvironmentBean.java

```diff
@@ -1225,4 +1236,127 @@ public void setNotifyCommitMarkableResourceRecoveryModuleOfCompleteBranches(
+        if(subordinateTransactionImporter == null && subordinateTransactionImporterClassName != null) {
+            synchronized(this) {
+                if(subordinateTransactionImporter == null && subordinateTransactionImporterClassName != null) {
+                    subordinateTransactionImporter = ClassloadingUtility.loadAndInstantiateClass(SubordinateTransactionImporter.class,
+                        subordinateTransactionImporterClassName, null);
+                }
+            }
+        }
+
+        return subordinateTransactionImporter;
+    }
+
+    /**
+     * Setting of class name that defines {@link UserTransactionOperationsProvider}.
+     * The provider is later used to get more info about {@link UserTransaction}.
+     *
+     * @param providerClassName  class name implementing {@link UserTransactionOperationsProvider}
+     */
+    public void setUserTransactionOperationsProviderClassName(String providerClassName) {
+        synchronized (this) {
+            if(providerClassName == null)
+            {
+                this.userTransactionOperationsProvider = null;
+            }
+            else if(!providerClassName.equals(this.userTransactionOperationsProviderClassName))
+            {
+                this.userTransactionOperationsProvider = null;
+            }
+            this.userTransactionOperationsProviderClassName = providerClassName;
+        }
+    }
+
+    /**
+     * Get class name that is used as {@link UserTransactionOperationsProvider}
+     *
+     * @return class name implementing {@link UserTransactionOperationsProvider}
+     */
+    public String getUserTransactionOperationsProviderClassName() {
+        return this.userTransactionOperationsProviderClassName;
+    }
+
+    public void setUserTransactionOperationsProvider(UserTransactionOperationsProvider provider) {
+        UserTransactionOperationsProvider oldProvider = this.userTransactionOperationsProvider;
+        this.userTransactionOperationsProvider = provider;
+
+        if(provider == null)
+        {
+            this.userTransactionOperationsProviderClassName = null;
+        }
+        else if (provider != oldProvider)
+        {
+            this.userTransactionOperationsProviderClassName =
+                ClassloadingUtility.getNameForClass(provider);
+        }
+        this.userTransactionOperationsProvider = provider;
+    }
+
+    /**
+     * Returning singleton instance of {@link UserTransactionOperationsProvider} instantiated based
+     * based on name specified by {@link #setUserTransactionOperationsProviderClassName(String)}.<br>
+     * When class name is redefined during runtime there should be instantiated new provider.
+     *
+     * @return instance of class implementing {@link UserTransactionOperationsProvider}
+     */
+    public UserTransactionOperationsProvider getUserTransactionOperationsProvider() {
+
+        if(userTransactionOperationsProvider == null && userTransactionOperationsProviderClassName != null) {
```

Comment:
From my research Ondra is using the same pattern as on the other setter/getters.